src/mlpack/methods/reinforcement_learning/sac.hpp

```diff
@@ -0,0 +1,178 @@
+/**
+ * @file methods/reinforcement_learning/sac.hpp
+ * @author Nishant Kumar
+ *
+ * This file is the definition of SAC class, which implements the
+ * soft actor-critic algorithm.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_SAC_HPP
+#define MLPACK_METHODS_RL_SAC_HPP
+
+#include <mlpack/prereqs.hpp>
+
+#include "replay/random_replay.hpp"
+#include <mlpack/methods/ann/activation_functions/tanh_function.hpp>
+#include <mlpack/methods/ann/loss_functions/mean_squared_error.hpp>
+#include <mlpack/methods/ann/visitor/parameters_visitor.hpp>
+#include "training_config.hpp"
+
+namespace mlpack {
+namespace rl {
+
+/**
+ * TODO: Add citation
+ *
+ * @tparam EnvironmentType The environment of the reinforcement learning task.
+ * @tparam NetworkType The network to compute action value.
+ * @tparam UpdaterType How to apply gradients when training.
+ * @tparam ReplayType Experience replay method.
+ */
+template <
+  typename EnvironmentType,
+  typename QNetworkType,
+  typename PolicyNetworkType,
+  typename UpdaterType,
+  typename ReplayType = RandomReplay<EnvironmentType>
+>
+class SAC
+{
+ public:
+  //! Convenient typedef for state.
+  using StateType = typename EnvironmentType::State;
+
+  //! Convenient typedef for action.
+  using ActionType = typename EnvironmentType::Action;
+
+  /**
+   * Create the SAC object with given settings.
+   *
+   * If you want to pass in a parameter and discard the original parameter
+   * object, be sure to use std::move to avoid unnecessary copy.
+   *
+   * @param config Hyper-parameters for training.
+   * @param learningQ1Network The network to compute action value.
+   * @param policyNetwork The network to produce an action given a state.
+   * @param replayMethod Experience replay method.
+   * @param qNetworkUpdater How to apply gradients to Q network when training.
+   * @param policyNetworkUpdater How to apply gradients to policy network
+   *        when training.
+   * @param environment Reinforcement learning task.
+   */
+  SAC(TrainingConfig& config,
+      QNetworkType& learningQ1Network,
+      PolicyNetworkType& policyNetwork,
+      ReplayType& replayMethod,
+      UpdaterType qNetworkUpdater = UpdaterType(),
+      UpdaterType policyNetworkUpdater = UpdaterType(),
+      EnvironmentType environment = EnvironmentType());
+
+  /**
+    * Clean memory.
+    */
+  ~SAC();
+
+  /**
+   * Softly update the learning Q_network parameters to the target Q_network
```

Comment:
```suggestion
   * Softly update the learning Q network parameters to the target Q_network
```

Couple other places to change this